File: kumi/sdk/csharp/EnvParser.cs
namespace Kumi;

/// <summary>
/// Simple .env file parser. Loaded values are stored in a static dictionary
/// and are accessible via <see cref="GetEnv"/>. Existing environment variables
/// are never overwritten.
/// </summary>
internal static class EnvParser
{
    private static readonly Dictionary<string, string> _loaded = new();

    public static void LoadEnvFile(string filePath)
    {
        if (!File.Exists(filePath)) return;

        foreach (var rawLine in File.ReadAllLines(filePath))
        {
            var line = rawLine.Trim();
            if (string.IsNullOrEmpty(line) || line.StartsWith('#'))
                continue;

            var eq = line.IndexOf('=');
            if (eq < 0) continue;

            var key = line[..eq].Trim();
            var value = line[(eq + 1)..].Trim();

            if (value.Length >= 2)
            {
                if ((value.StartsWith('"') && value.EndsWith('"')) ||
                    (value.StartsWith('\'') && value.EndsWith('\'')))
                {
                    value = value[1..^1];
                }
            }

            var existing = GetEnv(key);
            if (string.IsNullOrEmpty(existing))
                _loaded[key] = value;
        }
    }

    public static string? GetEnv(string key)
    {
        var env = Environment.GetEnvironmentVariable(key);
        if (!string.IsNullOrEmpty(env)) return env;
        return _loaded.TryGetValue(key, out var v) ? v : null;
    }

    public static string GetEnv(string key, string fallback)
    {
        var v = GetEnv(key);
        return !string.IsNullOrEmpty(v) ? v : fallback;
    }
}

File: kumi/sdk/java/src/main/java/io/kumi/EnvParser.java
package io.kumi;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Simple .env file parser. Sets values into System properties (not
 * System.getenv which is unmodifiable) unless the env var already exists.
 *
 * <p>Values are accessible via {@link EnvParser#getEnv(String)} which
 * checks both {@code System.getenv} and {@code System.getProperty}.</p>
 */
public final class EnvParser {
    private EnvParser() {}

    /**
     * Load a .env file. Existing environment variables and previously
     * loaded values are never overwritten.
     */
    public static void loadEnvFile(String filePath) {
        Path path = Path.of(filePath);
        if (!Files.isRegularFile(path)) return;

        try (BufferedReader reader = Files.newBufferedReader(path)) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) continue;
                int eq = line.indexOf('=');
                if (eq < 0) continue;
                String key = line.substring(0, eq).trim();
                String value = line.substring(eq + 1).trim();
                if (value.length() >= 2) {
                    if ((value.startsWith("\"") && value.endsWith("\"")) ||
                        (value.startsWith("'") && value.endsWith("'"))) {
                        value = value.substring(1, value.length() - 1);
                    }
                }
                if (getEnv(key) == null || getEnv(key).isEmpty()) {
                    System.setProperty(key, value);
                }
            }
        } catch (IOException ignored) {
        }
    }

    /**
     * Read a configuration value: checks System.getenv first, then
     * System.getProperty (for values loaded from .env).
     */
    public static String getEnv(String key) {
        String v = System.getenv(key);
        if (v != null && !v.isEmpty()) return v;
        return System.getProperty(key);
    }

    public static String getEnv(String key, String fallback) {
        String v = getEnv(key);
        return (v != null && !v.isEmpty()) ? v : fallback;
    }
}
